react/javascript/demo-external-api/src/components/SheetsForm.jsx

```diff
@@ -0,0 +1,460 @@
+    if (showErrors) {
+      setValidationErrors(errors)
+    }
+    return Object.keys(errors).length === 0
+  }
+
+  const newStateOptions = useMemo(() => {
+    if (stateFilter === '') return stateOptions
+    return stateOptions.filter((option) =>
+      option.label.toLowerCase().includes(stateFilter.toLowerCase())
+    )
+  }, [stateFilter])
+
+  const onSubmit = (e) => {
+    e.preventDefault()
+    if (validate(true)) {
+      onRowSave(index, [name, gender, year, state, major, activity])
+      closeDialog()
+    }
+  }
+
+  const closeDialog = () => {
+    setValidationErrors({})
+    setShowForm(false)
+    setCurrentIndex(undefined)
+    setName('')
+    setGender('')
+    setYear('')
+    setState('')
+    setMajor('')
+    setActivity('')
+  }
+
+  return (
+    <Dialog isOpen={showForm} onClose={() => closeDialog()}>
+      <DialogLayout header="Row details">
+        <Form onSubmit={onSubmit} validationMessages={validationErrors}>
+          <FieldText
+            label="Name"
+            name="name"
+            value={name}
+            onChange={(e) => setName(e.currentTarget.value)}
+          />
+          <FieldSelect
+            label="Gender"
+            value={gender}
+            name="gender"
+            options={[
+              { value: 'Male' },
+              { value: 'Female' },
+              { value: 'Other' },
+            ]}
+            onChange={(value) => setGender(value)}
+          />
+          <FieldSelect
+            label="Year"
+            value={year}
+            name="year"
+            options={[
+              { value: '1. Freshman' },
+              { value: '2. Sophomore' },
+              { value: '3. Junior' },
+              { value: '4. Senior' },
+            ]}
+            onChange={(value) => setYear(value)}
+          />
+          <FieldSelect
+            label="State"
+            value={state}
+            name="state"
+            isFilterable={true}
+            onFilter={(value) => setStateFilter(value)}
+            onChange={(value) => setState(value)}
+            options={newStateOptions}
+          />
+          <FieldSelect
+            label="Major"
+            value={major}
+            name="major"
+            onChange={(value) => setMajor(value)}
+            options={majors}
+          />
+          <FieldSelect
+            label="Activity"
+            value={activity}
+            name="activity"
+            onChange={(value) => setActivity(value)}
+            options={activities}
+            isClearable={true}
+          />
+          <Button type="submit">{isInsert ? 'Insert' : 'Update'}</Button>
+        </Form>
+      </DialogLayout>
+    </Dialog>
+  )
+}
+
+SheetsForm.propTypes = {
+  showForm: PropTypes.bool,
+  setShowForm: PropTypes.func.isRequired,
+  row: PropTypes.array,
+  index: PropTypes.number,
+  isInsert: PropTypes.bool,
+  onRowSave: PropTypes.func.isRequired,
+  rows: PropTypes.array.isRequired,
+}
+
+const stateOptions = [
```

Comment:
Could we add a json loader to sort of separate out concerns?

Reply:
Done